free/src/main/scala/cats/free/FreeT.scala

```diff
@@ -0,0 +1,263 @@
+        case Suspend(m) => M.map(m) {
+          case Xor.Left(a) => g1.f(a)
+          case Xor.Right(s) => liftF[S, M, g1.A](s).flatMap(g1.f)
+        }
+        case g0 @ FlatMapped(_, _) => g0.a.flatMap(g0.f(_).flatMap(g1.f)).toM
+      }
+    }
+
+  @tailrec
+  private def step: FreeT[S, M, A] =
+    this match {
+      case g @ FlatMapped(_, _) => g.a match {
+        case g0 @ FlatMapped(_, _) => g0.a.flatMap(a => g0.f(a).flatMap(g.f)).step
+        case _ => g
+      }
+      case x => x
+    }
+
+  override def toString(): String = "FreeT(...)"
+}
+
+object FreeT extends FreeTInstances {
+  /** Suspend the computation with the given suspension. */
+  private[free] case class Suspend[S[_], M[_], A](a: M[A Xor S[A]]) extends FreeT[S, M, A]
+
+  /** Call a subroutine and continue with the given function. */
+  private[free] case class FlatMapped[S[_], M[_], A0, B](a0: FreeT[S, M, A0], f0: A0 => FreeT[S, M, B]) extends FreeT[S, M, B] {
+    type A = A0
+    def a: FreeT[S, M, A] = a0
+    def f: A => FreeT[S, M, B] = f0
+  }
+
+  /** Return the given value in the free monad. */
+  def pure[S[_], M[_], A](value: A)(implicit M: Applicative[M]): FreeT[S, M, A] = Suspend(M.pure(Xor.left(value)))
+
+  def suspend[S[_], M[_], A](a: M[A Xor S[FreeT[S, M, A]]])(implicit M: Applicative[M]): FreeT[S, M, A] =
+    liftT(a).flatMap({
+      case Xor.Left(a) => pure(a)
+      case Xor.Right(s) => roll(s)
+    })
+
+  def tailRecM[S[_], M[_]: Applicative, A, B](a: A)(f: A => FreeT[S, M, A Xor B]): FreeT[S, M, B] =
+    f(a).flatMap {
+      case Xor.Left(a0) => tailRecM(a0)(f)
+      case Xor.Right(b) => pure[S, M, B](b)
+    }
+
+  def liftT[S[_], M[_], A](value: M[A])(implicit M: Functor[M]): FreeT[S, M, A] =
+    Suspend(M.map(value)(Xor.left))
+
+  /** A version of `liftT` that infers the nested type constructor. */
+  def liftTU[S[_], MA](value: MA)(implicit M: Unapply[Functor, MA]): FreeT[S, M.M, M.A] =
+    liftT[S, M.M, M.A](M.subst(value))(M.TC)
+
+  /** Suspends a value within a functor in a single step. Monadic unit for a higher-order monad. */
+  def liftF[S[_], M[_], A](value: S[A])(implicit M: Applicative[M]): FreeT[S, M, A] =
+    Suspend(M.pure(Xor.right(value)))
+
+  def roll[S[_], M[_], A](value: S[FreeT[S, M, A]])(implicit M: Applicative[M]): FreeT[S, M, A] =
+    liftF[S, M, FreeT[S, M, A]](value).flatMap(identity)
+
+}
+
+private[free] sealed trait FreeTInstances3 {
+  implicit def catsFreeMonadStateForFreeT[S[_], M[_], E](implicit M1: MonadState[M, E]): MonadState[FreeT[S, M, ?], E] =
+    new MonadState[FreeT[S, M, ?], E] with FreeTMonad[S, M] {
+      override def M = implicitly
+      override def get =
+        FreeT.liftT(M1.get)
+      override def set(s: E) =
+        FreeT.liftT(M1.set(s))
+    }
+}
+
+private[free] sealed trait FreeTInstances2 extends FreeTInstances3 {
+  implicit def catsFreeMonadErrorForFreeT[S[_], M[_]: RecursiveTailRecM, E](implicit E: MonadError[M, E]): MonadError[FreeT[S, M, ?], E] =
```

Comment:
Constrains the nested M to be also stack safe but if you think is not necessary we can remove this constrain.